README
WWW-Domain-Registry-Joker version 0.10
=========================================

The WWW-Domain-Registry-Joker module provides a Perl interface to
the Domain Management API (DMAPI) used by the Joker.com DNS registrar.
It is designed to help Joker.com resellers in automating the domain
registration and all the other relevant actions.

This distribution probably needs a bit more documentation :)

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Log::Dispatch
  LWP
  Test::More

AUTHOR

Peter Pentchev <roam@ringlet.net>

COPYRIGHT AND LICENCE

Copyright (C) 2007 - 2009 by Peter Pentchev

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

$CNsys: README 3201 2009-06-26 08:58:10Z roam $


Changes
Revision history for Perl extension WWW::Domain::Registry::Joker,
the Joker.com DMAPI interface module.

0.10	2009/06/26
	- README
	  - update the copyright years
	- WWW::Domain::Registry::Joker
	  - only send "!@!" for object modification requests; no need
	    to register contacts and domains with "!@!" in the empty
	    address parts
	  - update the copyright years

0.09	2008/11/14
	- Makefile.PL
	  - do not regenerate the META.yml each time, thus actually
	    making use of my updated version in 0.06!

0.08	2008/11/14
	- belatedly bump the copyright year on README, Makefile.PL,
	  and Joker.pm
	- WWW::Domain::Registry::Joker
	  - send "!@!" instead of an empty string, since the DMAPI
	    treats an empty string as "no change requested".
	    Hoster ticket #750

0.07	2008/10/17
	- add the META.yml file to the packing manifest

0.06	2008/10/17
	- add a META.yml file
	- Makefile.PL
	  - specify the Perl license for newer MakeMaker versions

0.05	2007/11/26
	- WWW::Domain::Registry::Joker
	  - change web hyperlinks to italics instead of POD links for
	    compatibility with Perl 5.6.x.
	- WWW::Domain::Registry::Joker::Response
	  - change web hyperlinks to italics instead of POD links for
	    compatibility with Perl 5.6.x.
	- t/WWW-Domain-Registry-Joker-podcheck.t
	  - fix the number of tests planned
	  - skip the num_warnings() test on early versions of Pod::Checker
	- bump module versions:
	  WWW::Domain::Registry::Joker::Response 0.04

0.04	2007/11/21
	- WWW::Domain::Registry::Joker::Loggish
	  - rename the err() method to error() to avoid clashes with
	    my own convention of an err() getter/setter method in
	    other classes
	  - append the newline in the log() method, not in debug() or
	    error()
	  - fix the error() POD documentation syntax
	- scripts/test.pl
	  - add usage info, thereby eliminating the last FIXME item
	    left in the code :)
	- t/WWW-Domain-Registry-Joker-coverage.t
	  - add a new test file running Test::Pod::Coverage tests
	- t/WWW-Domain-Registry-Joker-podcheck.t
	  - add a new test file running Pod::Checker tests

0.03	2007/11/14
	- fix the POD syntax of the Joker::Response class
	- remove the hooks for fake response parsing, always use
	  a real HTTP::Response object
	- properly check if the response passed in is indeed an object
	  before invoking OO methods on it
	- properly parse multiline response headers - they appear as
	  "duplicate" headers, so keep the old value, append the new one
	- fully perl-document the Joker::Response class
	- implement Joker->query_domain_list() properly
	- remove the Joker->checkavailability() misintended stub
	- remove the method prototypes in the Joker class, since they
	  serve absolutely no purpose in OOP classes
	- fully perl-document the WWW::Domain::Registry::Joker class
	- bump module versions:
	  WWW::Domain::Registry::Joker::Response 0.03

0.02	2007/11/09
	- remove a leftover copyright statement from Joker.pm.
	- add some perldoc documentation
	- Joker.pm checks if the lwp() parameter is a LWP::UserAgent
	- bump module versions:
	  WWW::Domain::Registry::Joker::Loggish 0.02
	  WWW::Domain::Registry::Joker::Response 0.02

0.01	2007/11/07
	- Initial semi-public release.

$CNsys: Changes 3202 2009-06-26 09:00:31Z roam $


scripts/tests.pl
#!/usr/bin/perl -w

# $CNsys: scripts/tests.pl 2375 2008-10-17 10:25:08Z roam $

# Copyright (C) 2007 by Peter Pentchev
#
# This library is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself, either Perl version 5.8.8 or,
# at your option, any later version of Perl 5 you may have available.

use strict;

use Getopt::Std;

use WWW::Domain::Registry::Joker;

MAIN:
{
	my ($l, $r, $x);
	my (%opts, %h);

	getopts('d', \%opts) or
	    die("Usage: tests.pl [-d]\n\t-d\tdisplay diagnostic messages.\n");
	$r = new WWW::Domain::Registry::Joker(
	    'username' => 'your-joker-username@example.com',
	    'password' => 'somekindasecret', 'debug' => $opts{'d'},
	    'fake' => 0,
	);

	if (0) {
	%h = $r->query_domain_list('');
	print "====== query-domain-list results:\n".
	    join('', map "- domain $_ expires $h{$_}->{exp}\n", sort keys %h).
	    "======\n";
	}

	if (0) {
	print $r->do_request('contact-create', 'tld' => 'net',
	    'fname' => 'Peter', 'lname' => 'Pentchev', 'individual' => 'Y',
	    'email' => 'roam@hoster.bg',
	    'address-1' => '37 Dobri Vojnikov Street',
	    'city' => 'Sofia', 'postal-code' => '1000', 'country' => 'BG',
	    'phone' => '+359-88-883-1313')."\n";
	}

	if (0) {
	print $r->do_request('ns-create', 'Host' => 'a.ns.ringringlet.net',
	    'IP' => '77.77.142.9')."\n";
	}

	if (0) {
	print $r->do_request('domain-register', 'domain' => 'ringringlet.net',
	    'period' => 24, 'status' => 'production',
	    'owner-c' => 'CNET-673793',
	    'billing-c' => 'CNET-673793',
	    'admin-c' => 'CNET-673793',
	    'tech-c' => 'CNET-673793',
	    'ns-list' => 'a.ns.ringlet.net:b.ns.ringlet.net')."\n";
	}

	if (0) {
	print $r->do_request('domain-delete', 'domain' => 'villa-straylight.info'
	    )."\n";
	}

	if (0) {
	%h = $r->result_list();
	print "Result list:\n".join("\n", map { join ' ', 
	    @{$_}{qw/tstamp svtrid procid reqtype reqobject status cltrid/}}
	    sort { $a->{'procid'} cmp $b->{'procid'} } values %h)."\n";
	}
}


Makefile.PL
use 5.006;

# $CNsys: Makefile.PL 2526 2008-11-14 11:22:46Z roam $

# Copyright (C) 2007, 2008 by Peter Pentchev
#
# This library is free software; you can redistribute it and/or modify
# it under the same terms as Perl itself, either Perl version 5.8.8 or,
# at your option, any later version of Perl 5 you may have available.

use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'WWW::Domain::Registry::Joker',
    VERSION_FROM      => 'lib/WWW/Domain/Registry/Joker.pm',
    PREREQ_PM         => {
	'Log::Dispatch'	=> 2,
	'LWP'		=> 5,
	'Test::More'	=> 0,
       	},
    NO_META           => 1,
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/WWW/Domain/Registry/Joker.pm',
       AUTHOR         => 'Peter Pentchev <roam@ringlet.net>') : ()),
    ($ExtUtils::MakeMaker::VERSION >= 6.3002 ? (LICENSE => "perl") : ()),
);


lib/WWW/Domain/Registry/Joker/Loggish.pm
package WWW::Domain::Registry::Joker::Loggish;

# $CNsys: lib/WWW/Domain/Registry/Joker/Loggish.pm 2375 2008-10-17 10:25:08Z roam $

use 5.006;
use strict;
use warnings;

use Log::Dispatch;
use Log::Dispatch::Screen;

our $VERSION = '0.02';

=head1 NAME

WWW::Domain::Registry::Joker::Loggish - a simple logging helper

=head1 SYNOPSIS

  use WWW::Domain::Registry::Joker::Loggish;

  @ISA = qw/WWW::Domain::Registry::Joker::Loggish/;

  $self = WWW::Domain::Registry::Joker::Loggish::new($proto,
  	'debug' => 1);

  $self->debug('Diagnostics for the masses');
  $self->error('Nobody loves me!');

  $self->log('notice', 'Something important is about to happen');
  $self->log('debug', 'Nobody ever bothers to read those...');

=head1 DESCRIPTION

The C<WWW::Domain::Registry::Joker::Loggish> class provides a simple
logging interface, implemented using the C<Log::Dispatch> module.
It is meant to serve as a parent class providing the C<log()>, C<debug()>,
and C<error()> methods so that other classes do not have to worry about
implementing them.

=head1 METHODS

The C<WWW::Domain::Registry::Joker::Loggish> class provides the following
methods:

=over 4

=item * new ( PARAMS )

Create a new C<WWW::Domain::Registry::Joker::Loggish> object with
the specified parameters:

=over 4

=item * debug

A boolean flag for the output of diagnostic messages - should the C<debug()>
method actually display the message passed or simply ignore it.

=item * log

The C<Log::Dispatch> object to use for the output; if not passed,
a new object will be created at first use - see the C<logger()> method
below.

=back

=cut

sub new($ %)
{
	my ($proto, %param) = @_;
	my $class = ref $proto || $proto;
	my $self;

	$self = {
		'debug'		=> 0,
		'log'		=> undef,
		%param,
	};
	bless $self, $class;
	return $self;
}

=item * logger ( [OBJECT] )

Get or set the C<Log::Dispatch> object used for the actual logging.

If no object is specified and no logging object has been set yet, this
method will create a C<Log::Dispatch> object and a C<Log::Dispatch::Screen>
destination set to output to the standard error stream.  If tis does not
suit the needs of the application, it should invoke the C<logger()> method
and pass its own C<Log::Dispatch> handler.  This may also be done at
object creation time by passing the C<log> parameter to the C<new()>
method.

=cut

sub logger($ $)
{
	my ($self, $log) = @_;

	if (defined($log)) {
		if (index(ref($log), '::') == -1 ||
		    !$log->isa('Log::Dispatch')) {
			die("Not a Log::Dispatch object: '".ref($log)."'");
		}
		$self->{'log'} = $log;
	} elsif (!defined($self->{'log'})) {
		$log = new Log::Dispatch();
		$log->add(new Log::Dispatch::Screen('name' => 'STDERR',
		    'min_level' => ($self->{'debug'}? 'debug': 'info')));
		$self->{'log'} = $log;
	}
	return $self->{'log'};
}

=item * log ( LEVEL, MESSAGE )

Log the specified message at the specified level.

This method invokes the C<logger()> method, so that a C<Log::Dispatch>
object will be created automatically at first use if none has been
specified.

=cut

sub log($ $ $)
{
	my ($self, $level, $msg) = @_;

	$self->logger()->log('level' => $level, 'message' => "$msg\n");
}

=item * debug ( MESSAGE )

Log a message with a priority of 'debug' using the C<log()> method.
Note that whether the message will actually be logged or not will
depend on the setting of the C<debug> property at object creation time.

=cut

sub debug($ $)
{
	my ($self, $msg) = @_;

	$self->log('debug', $msg);
}

=item * error ( MESSAGE )

Log a message with a priority of 'error' using the C<log()> method.

=cut

sub error($ $)
{
	my ($self, $msg) = @_;

	$self->log('error', $msg);
}

=back

=head1 SEE ALSO

L<Log::Dispatch>

=head1 BUGS

=over 4

=item *

Maybe there ought to be a way to toggle the display of diagnostic
messages after the object has been created.

=back

=head1 HISTORY

The C<WWW::Domain::Registry::Joker::Loggish> class was written by
Peter Pentchev in 2007.

=head1 AUTHOR

Peter Pentchev, E<lt>roam@ringlet.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by Peter Pentchev

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;


lib/WWW/Domain/Registry/Joker/Response.pm
package WWW::Domain::Registry::Joker::Response;

# $CNsys: lib/WWW/Domain/Registry/Joker/Response.pm 2375 2008-10-17 10:25:08Z roam $

use 5.006;
use strict;
use warnings;

use WWW::Domain::Registry::Joker::Loggish;

our @ISA = qw(WWW::Domain::Registry::Joker::Loggish);

our $VERSION = '0.04';

=head1 NAME

WWW::Domain::Registry::Joker::Response - parse a DMAPI response

=head1 SYNOPSIS

  use WWW::Domain::Registry::Joker::Response;

  $r = new WWW::Domain::Registry::Joker::Response();
  $r->parse($resp);
  print "$r->{Proc-Id}: $r->{'Status-Code'} $r->{Status-Text}\n";

=head1 DESCRIPTION

The C<WWW::Domain::Registry::Joker::Response> class is a helper parser
for the HTTP responses returned by the Joker.com DMAPI.  It examines
a response object, extracts the status and error flags, codes, and
descriptive messages, and makes them available as Perl object members.

=head1 METHODS

The C<WWW::Domain::Registry::Joker::Response> class defines the
following methods:

=over 4

=item new ()

Initialize a C<WWW::Domain::Registry::Joker::Response> object.
No user-serviceable parameters inside.

=cut

sub new($ %)
{
	my ($proto, %param) = @_;
	my $self;

	$self = WWW::Domain::Registry::Joker::Loggish::new($proto,
		'code'		=> 1337,
		'msg'		=> '(no status text)',
		'status'	=> '(no status line)',
		'success'	=> 0,
		%param,
	);
	return $self;
}

=item parse ( RESPONSE )

Parse a C<HTTP::Response> from the DMAPI and store the result code,
message, error, etc. into the respective fields of the object.
In addition to the C<code>, C<msg>, C<status>, and C<success> members
described above, the C<parse()> method may also set the C<Version>,
C<Proc-Id>, C<Status-Text>, C<Status-Code>, C<Error>, and any other
result description members as listed in the DMAPI specification.

=cut

sub parse($ $)
{
	my ($self, $resp) = @_;
	my ($in_data, $var, $val);
	my (@r, @data);

	die("No response object passed to ".ref($self)."->parse()\n")
	    unless defined($resp);
	$self->debug("parsing a response - '".ref($resp)."'");
	if (index(ref($resp), '::') == -1 || !$resp->isa('HTTP::Response')) {
		die("Not a HTTP response object in ".ref($self)."->parse()\n");
	}

	$self->{'success'} = $resp->is_success();
	$self->{'status'} = $resp->status_line();
	undef $self->{'Status-Code'};
	undef $self->{'Status-Text'};
	@r = split("\n", $resp->content());
	$self->debug(scalar(@r)." content lines");
	foreach (@r) {
		s/[\r\n]+$//;
		if ($in_data) {
			$self->debug("- data line $_");
			push @data, $_;
			next;
		}
		$self->debug("- line $_");
		if ($_ eq '') {
			$self->debug("- - end of header");
			$in_data = 1;
			next;
		} elsif (!/^([\w-]+):\s+(.*)$/) {
			$self->debug("- - bad format!");
			next;
		}
		($var, $val) = ($1, $2);
		if (defined($self->{$var})) {
			$self->{$var} .= " / $val";
		} else {
			$self->{$var} = $val;
		}
	}
	$self->{'data'} = [ @data ];
	if (defined($self->{'Status-Code'})) {
		$self->{'code'} = $self->{'Status-Code'};
	} else {
		$self->{'code'} = 1337;
	}
	if (defined($self->{'Status-Text'})) {
		$self->{'msg'} = $self->{'Status-Text'};
	} else {
		$self->{'msg'} = '(no status text)';
	}
	$self->debug("=== DMAPI response: is_success $self->{success} ".
	    "status line '$self->{status}', status code: $self->{code}, ".
	    "status text: '$self->{msg}', ".
	    "data lines ".scalar(@{$self->{'data'}})."\n");
	return 1;
}

=back

=head1 EXAMPLES

Create an object and parse an HTTP response:

  $r = new WWW::Domain::Registry::Joker::Response();
  eval {
  	$r->parse($resp);
  };
  if ($@) {
  	print STDERR "Could not parse the DMAPI response: $@\n";
  } elsif (!$r->{'success'}) {
  	print STDERR "DMAPI error: code $r->{code}, text $r->{msg}\n";
  	print STDERR "DMAPI error message: $r->{Error}\n"
  	    if $r->{'Error'};
  } else {
  	print "Successful DMAPI request: $r->{code} $r->{msg}\n";
  	print "Tracking process ID: $r->{Proc-ID}\n" if $r->{'Proc-ID'};
  }

=head1 ERRORS

The C<parse()> method will die on invalid input:

=over 4

=item *

no response parameter passed in;

=item *

the response parameter was not an C<HTTP::Response> or compatible object.

=back

If the response object is a valid DMAPI response, its C<success>, C<code>,
C<msg>, C<Error>, and other attributes are exposed as members of
the C<WWW::Domain::Registry::Joker::Response> object as shown above.

=head1 SEE ALSO

L<WWW::Domain::Registry::Joker>, L<HTTP::Response>

I<https://joker.com/faq/category/39/22-dmapi.html> - the Joker.com DMAPI
documentation

=head1 BUGS

None known so far ;)

=head1 HISTORY

The C<WWW::Domain::Registry::Joker::Response> class was written by
Peter Pentchev in 2007.

=head1 AUTHOR

Peter Pentchev, E<lt>roam@ringlet.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 by Peter Pentchev

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;


lib/WWW/Domain/Registry/Joker.pm
package WWW::Domain::Registry::Joker;

# $CNsys: lib/WWW/Domain/Registry/Joker.pm 3201 2009-06-26 08:58:10Z roam $

use 5.006;
use strict;
use warnings;

use LWP;
use URI::Escape;

use WWW::Domain::Registry::Joker::Loggish;

use WWW::Domain::Registry::Joker::Response;

our @ISA = qw(WWW::Domain::Registry::Joker::Loggish);

our $VERSION = '0.10';

=head1 NAME

WWW::Domain::Registry::Joker - an interface to the Joker.com DMAPI

=head1 SYNOPSIS

  use WWW::Domain::Registry::Joker;

  $reg = new WWW::Domain::Registry::Joker('username' => 'testuser',
    'password' => 'secret', 'debug' => 1);

  @res = $reg->result_list();

  eval {
    $procid = $reg->do_request('ns-create', 'Host' => 'a.ns.example.com',
      'IP' => '192.168.13.1');
  };
  if ($@) {
    warn("Joker request failed: $@\n");
  }

=head1 DESCRIPTION

The C<WWW::Domain::Registry::Joker> module provides a Perl interface to
the Domain Management API (DMAPI) used by the Joker.com DNS registrar.
It is designed to help Joker.com resellers in automating the domain
registration and all the other relevant actions.

The recommended usage of the C<WWW::Domain::Registry::Joker> class is
to create an object, initialize it with the Joker.com reseller's username
and password, and then use it to send all the DMAPI requests.  This will
take care of caching both login credentials and network connections
(at least as far as C<LWP> takes care of caching connections to the same
webserver).

In most cases it is not necessary to invoke the C<login()> method
explicitly, since all the "real" action methods check for an authentication
token and invoke C<login()> if there is none yet.

=head1 METHODS

The C<WWW::Domain::Registry::Joker> class defines the following methods:

=over 4

=item * new ( PARAMS )

Create a new Joker.com interface object with the specified parameters:

=over 4

=item * username

The Joker.com reseller authentication username.

=item * password

The Joker.com reseller authentication password.

=item * debug

The diagnostic output level, 0 for no diagnostic messages.

=item * dmapi_url

The URL to use for Joker.com Domain Management API (DMAPI) requests;
if not specified, the standard URL I<https://dmapi.joker.com/request>
is used.

=back

=cut

sub new($ %)
{
	my ($proto, %param) = @_;
	my $class = ref $proto || $proto;
	my $self;

	$self = WWW::Domain::Registry::Joker::Loggish::new($proto,
		'authsid' => undef,
		'lwp' => undef,
		'err' => undef,
		'fake' => 0,
		'password' => undef,
		'dmapi_url' => 'https://dmapi.joker.com/request',
		'username' => undef,
		%param,
	);
	return $self;
}

=item * lwp ( [OBJECT] )

Get or set the C<LWP::UserAgent> object used for sending the actual
requests to the Joker.com web API.

This method should probably never interest any consumers of this class :)

=cut

sub lwp($ $)
{
	my ($self, $obj) = @_;

	if (defined($obj)) {
		if (index(ref($obj), '::') == -1 ||
		    !$obj->isa('LWP::UserAgent')) {
			die("WWW::Domain::Registry::Joker->lwp() requires a ".
			    "LWP::UserAgent object, not '".ref($obj)."'\n");
		}
		$self->{'lwp'} = $obj;
	} elsif (!defined($self->{'lwp'})) {
		$obj = new LWP::UserAgent();
		$self->{'lwp'} = $obj;
	}
	return $self->{'lwp'};
}

=item * build_request ( REQUEST, PARAMS )

Build a C<HTTP::Request> object for submitting an actual request to
the Joker.com API.

This method should probably never interest any consumers of this class :)

=cut

sub build_request($ $ %)
{
	my ($self, $req, %param) = @_;
	my ($s, $k, $v, $l, $i);

	$s = "$self->{'dmapi_url'}/$req";
	if (%param) {
		$s .= '?'.join '&', map
		    uri_escape($_).'='.uri_escape($param{$_}), keys %param;
	}
	return new HTTP::Request('GET' => $s);
}

=item * login ()

Send a DMAPI login authentication request and obtain the auth SID for
use in the follow-up actual requests.  The I<username> and I<password>
member variables must be initialized.

=cut

sub login($)
{
	my ($self) = @_;
	my ($req, $hresp, $resp);

	die("No DMAPI credentials supplied")
	    unless defined($self->{'username'}) && defined($self->{'password'});
	$req = $self->build_request('login', 'username' => $self->{'username'},
	    'password' => $self->{'password'});
	$self->debug("=== DMAPI login request\n".$req->as_string()."\n===\n");
	$hresp = $self->lwp()->request($req);
	$resp = new WWW::Domain::Registry::Joker::Response(
	    'debug' => $self->{'debug'}, 'log' => $self->{'log'});
	$resp->parse($hresp);
	die("DMAPI login: $resp->{Error}\n") if defined($resp->{'Error'});
	die("DMAPI login: $resp->{status}\n") unless $resp->{'success'};
	die("DMAPI login error: $resp->{code} $resp->{msg}\n")
	    unless defined($resp->{'code'}) && $resp->{'code'} == 0;
	die("DMAPI login parse error: no auth session ID\n")
	    unless defined($resp->{'Auth-Sid'});
	$self->{'authsid'} = $resp->{'Auth-Sid'};
	return $self->{'authsid'};
}

=item * query_domain_list ( PATTERN )

Return information about the domains registered by this reseller whose names
match the supplied pattern.  Returns a hash indexed by domain name, each
element of which is a hash:

=over 4

=item * domain

The domain name (yes, again :))

=item * exp

The expiration date of the domain registration.

=back

Invokes the C<login()> method if necessary.

=cut

sub query_domain_list($ $)
{
	my ($self, $pattern) = @_;
	my ($req, $hresp, $resp);
	my (%res);

	if (!defined($self->{'authsid'})) {
		return undef unless $self->login();
	}
	$req = $self->build_request('query-domain-list',
	    'Auth-Sid' => $self->{'authsid'},
	    'pattern' => $pattern);
	$self->debug("=== DMAPI qdlist request\n".$req->as_string()."\n===\n");
	$hresp = $self->lwp()->request($req);
	$resp = new WWW::Domain::Registry::Joker::Response(
	    'debug' => $self->{'debug'}, 'log' => $self->{'log'});
	$resp->parse($hresp);
	die("DMAPI qdlist: $resp->{Error}\n") if defined($resp->{'Error'});
	die("DMAPI qdlist: $resp->{status}\n") unless $resp->{'success'};
	die("DMAPI qdlist error: $resp->{code} $resp->{msg}\n")
	    unless defined($resp->{'code'}) && $resp->{'code'} == 0;
	foreach (@{$resp->{'data'}}) {
		if (!/^(\S+)\s+(\S+)$/) {
			$self->debug("- invalid format $_");
			next;
		}
		$res{$1} = { 'domain' => $1, 'exp' => $2 };
	}
	return %res;
}

=item * do_request ( REQUEST, PARAMS )

Send a DMAPI request with the name specified in C<REQUEST> and
parameters in the C<PARAMS> hash.  The request name string and
the parameters (required and optional) are as specified by
the DMAPI documentation

Note that for object modification requests (those which type is
C<domain-owner-change> or ends in C<-modify>) if a parameter is supplied
with the empty string as a value, the C<do_request()> method will send
the I<"!@!"> string instead, since the DMAPI considers empty values to mean
no change requested.

Invokes the C<login()> method if necessary.

=cut

sub do_request($ $ %)
{
	my ($self, $type, %data) = @_;
	my ($req, $hresp, $resp);
	my (%d);

	if (!defined($self->{'authsid'})) {
		return undef unless $self->login();
	}
	foreach (keys %data) {
		if (defined($data{$_}) && length($data{$_})) {
			$d{$_} = $data{$_};
		} elsif ($type eq 'domain-owner-change' ||
		    $type =~ /-modify$/) {
			$d{$_} = '!@!';
		}
	}
	$req = $self->build_request("$type",
	    'Auth-Sid' => $self->{'authsid'},
	    %d);
	$self->debug("=== DMAPI $type request\n".$req->as_string()."\n===\n");
	$hresp = $self->lwp()->request($req);
	$resp = new WWW::Domain::Registry::Joker::Response(
	    'debug' => $self->{'debug'}, 'log' => $self->{'log'});
	$resp->parse($hresp);
	die("DMAPI $type: $resp->{Error}\n") if defined($resp->{'Error'});
	die("DMAPI $type: $resp->{status}\n") unless $resp->{'success'};
	die("DMAPI $type error: $resp->{code} $resp->{msg}\n")
	    unless defined($resp->{'code'}) && $resp->{'code'} == 0;
	die("DMAPI $type - no processing ID returned!\n")
	    unless defined($resp->{'Proc-ID'});
	return $resp->{'Proc-ID'};
}

=item * result_list ()

Obtain the list of processed requests from the Joker.com DMAPI and
the corresponding result status and object ID (where applicable).
Returns a hash indexed by DMAPI I<Proc-Id> values.

Invokes the C<login()> method if necessary.

=cut

sub result_list($)
{
	my ($self) = @_;
	my ($req, $hresp, $resp);
	my (@r);
	my (%a, %res);

	if (!defined($self->{'authsid'})) {
		return undef unless $self->login();
	}
	$req = $self->build_request('result-list',
	    'Auth-Sid' => $self->{'authsid'});
	$self->debug("=== DMAPI rlist request\n".$req->as_string()."\n===\n");
	$hresp = $self->lwp()->request($req);
	$resp = new WWW::Domain::Registry::Joker::Response(
	    'debug' => $self->{'debug'}, 'log' => $self->{'log'});
	$resp->parse($hresp);
	die("DMAPI rlist: $resp->{Error}\n") if defined($resp->{'Error'});
	die("DMAPI rlist: $resp->{status}\n") unless $resp->{'success'};
	die("DMAPI rlist error: $resp->{code} $resp->{msg}\n")
	    unless defined($resp->{'code'}) && $resp->{'code'} == 0;
	%res = ();
	foreach (@{$resp->{'data'}}) {
		@r = split /\s+/;
		if ($#r != 6) {
			$self->debug("Unrecognized result-list line: $_");
			next;
		}
		@a{qw/tstamp svtrid procid reqtype reqobject status cltrid/} =
		    @r;
		$self->debug("result $a{procid} $a{status} $a{reqobject}");
		$res{$a{'procid'}} = { %a };
	}
	return %res;
}

=back

=head1 EXAMPLES

Initialize a C<WWW::Domain::Registry::Joker> object with your reseller's
username and password:

  $jreq = new WWW::Domain::Registry::Joker('username' => 'me@example.com',
  	'password' => 'somekindofsecret');

Fetch the list of pending and processed requests and their status:

  %h = $jreq->result_list();
  foreach (sort { $a->{'procid'} cmp $b->{'procid'} } values %h) {
  	print join("\t",
  	    @{$_}{qw/tstamp svtrid procid reqtype reqobject status cltrid/}).
	    "\n";
  }

Register a new nameserver:

  eval {
  	$jreq->do_request('ns-create', 'Host' => 'a.ns.example.net',
  	    'IP' => '192.168.13.7');
  };
  print STDERR "ns-create error: $@\n" if ($@);

Maybe some more examples are needed here :)

=head1 ERRORS

All the user-invoked methods die on any Joker.com errors with
a suitable error message placed in $@.

=head1 SEE ALSO

I<https://joker.com/faq/category/39/22-dmapi.html> - the Joker.com DMAPI
documentation

=head1 BUGS

=over 4

=item *

Reorder the methods placing the user-serviceable ones first.

=item *

Move C<WWW::Domain::Registry::Loggish> to a separate distribution?

=item *

Better error handling; exceptions?  Error.pm?  Something completely
different?  Croak?

=item *

Croak instead of die here and there.

=back

=head1 HISTORY

The C<WWW::Domain::Registry::Joker> class was written by Peter Pentchev
in 2007.

=head1 AUTHOR

Peter Pentchev, E<lt>roam@ringlet.netE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2007 - 2009 by Peter Pentchev

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.8 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
